# Car Insurance Data Analysis

## Exploratory Data Analysis


**Exploratory Data Analysis (EDA)** is a crucial step in any data engineering and anlytics project to understand the dataset before performing further analysis. It involves summarizing key characteristics and visualizing data patterns, outliers, and trends.

**Step 1. Understand the Data Structure**
- **Data loading**: First, load the dataset and examine its structure.

- **Check the dimensions**: Find the number of rows and columns.

- **Look at the first few records**: Get a sample of the dataset.


In [1]:
# Import necessary libraries
import matplotlib.pyplot as plt
import seaborn as sn
import pandas as pd
import numpy as np
import os, sys
# Add the 'scripts' directory to the Python path for module imports
sys.path.append(os.path.abspath(os.path.join('..', 'scripts')))

Load the car insurance dataset

In [2]:
# Import load_data function from scripts
from load_data import load_data

# read the dataset 

data = load_data('../data.zip', filename='MachineLearningRating_v3.txt')

In [3]:
# Explore the first few rows
data.head()

UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
0               145249     12827  2015-03-01 00:00:00             True   
1               145249     12827  2015-05-01 00:00:00             True   
2               145249     12827  2015-07-01 00:00:00             True   
3               145255     12827  2015-05-01 00:00:00             True   
4               145255     12827  2015-07-01 00:00:00             True   

  Citizenship          LegalType Title Language                 Bank  \
0              Close Corporation    Mr  English  First National Bank   
1              Close Corporation    Mr  English  First National Bank   
2              Close Corporation    Mr  English  First National Bank   
3              Close Corporation    Mr  English  First National Bank   
4              Close Corporation    Mr  English  First National Bank   

       AccountType  ...                    ExcessSelected CoverCategory  \
0  Current account  ...             Mobility - Windscreen    Windscreen   
1  Current account  ...             Mobility - Windscreen    Windscreen   
2  Current account  ...             Mobility - Windscreen    Windscreen   
3  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   
4  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   

    CoverType            CoverGroup              Section  \
0  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
1  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
2  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
3  Own Damage  Comprehensive - Taxi  Motor Comprehensive   
4  Own Damage  Comprehensive - Taxi  Motor Comprehensive   

                           Product StatutoryClass StatutoryRiskType  \
0  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
1  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
2  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
3  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
4  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   

   TotalPremium TotalClaims  
0     21.929825         0.0  
1     21.929825         0.0  
2      0.000000         0.0  
3    512.848070         0.0  
4      0.000000         0.0  

[5 rows x 52 columns]

In [4]:
# Find the number of rows and columns
data.shape

(1000098, 52)

The dataset has 100,0098 rows and 52 columns

**Data Cleaning:**

- Handle missing or incomplete data: Since the dataset spans several categories, some fields may have missing values. 


In [5]:
# Import the Class to data processing

from data_processing import DataProcessing
# Create instance of the class
data_processing = DataProcessing(data)

# Summary of Missing data
missing_summary = data_processing.missing_data_summary()

# Display results
missing_summary

Missing Count  Percentage (%)
NumberOfVehiclesInFleet        1000098      100.000000
CrossBorder                     999400       99.930207
CustomValueEstimate             779642       77.956560
Converted                       641901       64.183810
Rebuilt                         641901       64.183810
WrittenOff                      641901       64.183810
NewVehicle                      153295       15.327998
Bank                            145961       14.594670
AccountType                      40232        4.022806
Gender                            9536        0.953507
MaritalStatus                     8259        0.825819
Cylinders                          552        0.055195
cubiccapacity                      552        0.055195
kilowatts                          552        0.055195
NumberOfDoors                      552        0.055195
VehicleIntroDate                   552        0.055195
Model                              552        0.055195
make                               552        0.055195
VehicleType                        552        0.055195
mmcode                             552        0.055195
bodytype                           552        0.055195
CapitalOutstanding                   2        0.000200

**Drop Columns with High Missing Data:**
- Columns with high missing values offer little analytical value.

**Dropped Columns**

- `NumberOfVehiclesInFleet` (100% missing)
- `CrossBorder` (~99.93%)
- `CustomValueEstimate` (~77.96%)
- `Converted, Rebuilt, WrittenOff` (~64.18%)


In [ ]:
cols_to_drop = ['NumberOfVehiclesInFleet', 
                'CrossBorder', 
                'CustomValueEstimate', 
                'Converted', 'Rebuilt', 
                'WrittenOff']

# Drop these columns
data = data_processing.handle_missing_data('high', cols_to_drop)


**Impute Moderate Missing Data:**
- Imputation preserves useful information, using the mode for categorical and median for numerical columns.

**Imputed Columns:**

- `NewVehicle` (~15.33%)
- `Bank` (~14.59%)
- `AccountType` (~4.02%)

In [7]:
# Impute or drop columns with moderate missing data
missing_cols = ['NewVehicle', 'Bank', 'AccountType']
data = data_processing.handle_missing_data('moderate', missing_cols)

**Handle Low Missing Data - Standard Imputation**

- These columns can be reasonably imputed without affecting data quality.

**Imputed Columns:**

- `Gender` (~0.95%)
- `MaritalStatus` (~0.83%)
- `Various vehicle-related columns (~0.055% each): `Cylinders, CubicCapacity, Kilowatts`, etc.
 


In [8]:
# Handle low missing data (standard imputation)
missing_cols = ['Gender', 'MaritalStatus', 'Cylinders', 'cubiccapacity', 
                'kilowatts', 'NumberOfDoors', 'VehicleIntroDate', 'Model', 
                'make', 'VehicleType', 'mmcode', 'bodytype', 'CapitalOutstanding']

data = data_processing.handle_missing_data('low', missing_cols)


**Overall Decision Summary:**

- High missing data: Dropped.

- Moderate missing data: Imputed with mode (categorical) or median (numerical).

- Low missing data: Imputed to avoid unnecessary data loss.

You can decide on appropriate methods (imputation, removal) based on the importance of each feature.
Convert categorical columns (e.g., MaritalStatus, VehicleType, Bank) into appropriate formats (e.g., categories, dummy variables) for analysis.

**Summarize Key Statistics**

- **Descriptive statistics:**
    - In the descriptive statistics calculate and examine the variability for numerical features such as TotalPremium, TotalClaim, etc.

- **Distribution of numerical features:** Understand the range, variance, and skewness of features.

In [9]:
# Statistic summary of numerical features
display(data.describe())

UnderwrittenCoverID      PolicyID    PostalCode        mmcode  \
count         1.000098e+06  1.000098e+06  1.000098e+06  1.000098e+06   
mean          1.048175e+05  7.956682e+03  3.020601e+03  5.488056e+07   
std           6.329371e+04  5.290039e+03  2.649854e+03  1.360059e+07   
min           1.000000e+00  1.400000e+01  1.000000e+00  4.041200e+06   
25%           5.514300e+04  4.500000e+03  8.270000e+02  6.005692e+07   
50%           9.408300e+04  7.071000e+03  2.000000e+03  6.005842e+07   
75%           1.391900e+05  1.107700e+04  4.180000e+03  6.005842e+07   
max           3.011750e+05  2.324600e+04  9.870000e+03  6.506535e+07   

       RegistrationYear     Cylinders  cubiccapacity     kilowatts  \
count      1.000098e+06  1.000098e+06   1.000098e+06  1.000098e+06   
mean       2.010225e+03  4.046616e+00   2.466869e+03  9.721553e+01   
std        3.261391e+00  2.939410e-01   4.427106e+02  1.939061e+01   
min        1.987000e+03  0.000000e+00   0.000000e+00  0.000000e+00   
25%        2.008000e+03  4.000000e+00   2.237000e+03  7.500000e+01   
50%        2.011000e+03  4.000000e+00   2.694000e+03  1.110000e+02   
75%        2.013000e+03  4.000000e+00   2.694000e+03  1.110000e+02   
max        2.015000e+03  1.000000e+01   1.288000e+04  3.090000e+02   

       NumberOfDoors    SumInsured  CalculatedPremiumPerTerm  TotalPremium  \
count   1.000098e+06  1.000098e+06              1.000098e+06  1.000098e+06   
mean    4.019239e+00  6.041727e+05              1.178757e+02  6.190550e+01   
std     4.681854e-01  1.508332e+06              3.997017e+02  2.302845e+02   
min     0.000000e+00  1.000000e-02              0.000000e+00 -7.825768e+02   
25%     4.000000e+00  5.000000e+03              3.224800e+00  0.000000e+00   
50%     4.000000e+00  7.500000e+03              8.436900e+00  2.178333e+00   
75%     4.000000e+00  2.500000e+05              9.000000e+01  2.192982e+01   
max     6.000000e+00  1.263620e+07              7.442217e+04  6.528260e+04   

        TotalClaims  
count  1.000098e+06  
mean   6.486119e+01  
std    2.384075e+03  
min   -1.200241e+04  
25%    0.000000e+00  
50%    0.000000e+00  
75%    0.000000e+00  
max    3.930921e+05